lib: switch to Number.isNaN

BridgeAR opened the pull request:
This enables the eslint no-self-compare rule. 

The reason is to switch to the more natural `Number.isNaN()`check. I benchmarked both versions and they are on par so we do not have to use those old ones anymore.

I also removed a `NaN` check that was obsolete and added a test for that.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
lib

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/13141/

lpinca commented on lib/buffer.js:
```diff
@@ -999,15 +997,17 @@ function adjustOffset(offset, length) {
-  } else {
-    return offset < length ? offset : length;
   }
+  if (Number.isNaN(offset)) {
```
Any reason for not merging this with the `offset === 0` case on line 1000?

BridgeAR replied:
`NaN` is just a safety check. The other cases are not. Therefore I prioritized the other case. Due to your comment I actually saw that it is possible to move the NaN check down to be the last.

BridgeAR commented:
I realized that another NaN check was obsolete, so I removed it as well and moved a NaN check further down. I also changed all `isNaN` to `Number.isNaN`. This is fine since we always had numbers in those cases.

PTAL

richardlau commented on test/parallel/test-buffer-arraybuffer.js:
```diff
@@ -148,3 +148,6 @@ b.writeDoubleBE(11.11, 0, true);
   });
 }
+
+// Test a array like entry with the length set to NaN.
```
`an array-like`?

BridgeAR commented:
New CI before landing https://ci.nodejs.org/job/node-test-pull-request/13205/

(I am going to fix the typo when landing).

BridgeAR commented:
Landed in 43b8ce4ce741a8a6a435e4073ca5a9ec188d4515, 96c57fbfaa8c1644648fa1414c9ad131b1320b60.

I fixed the typo accordingly.
